src/main/java/org/jenkinsci/plugins/workflow/job/console/NewNodeConsoleNote.java

```diff
@@ -0,0 +1,131 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.job.console;
+
+import com.google.common.base.Predicates;
+import hudson.Extension;
+import hudson.MarkupText;
+import hudson.Util;
+import hudson.console.ConsoleAnnotationDescriptor;
+import hudson.console.ConsoleAnnotator;
+import hudson.console.ConsoleNote;
+import hudson.model.TaskListener;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.graphanalysis.Filterator;
+import org.jenkinsci.plugins.workflow.graphanalysis.FlowScanningUtils;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.support.actions.AnnotatedLogAction;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Console line with note printed when a new {@link FlowNode} is added to the graph.
+ * Defines the {@code pipeline-new-node} CSS class and several attributes which may be used to control subsequent behavior:
+ * <ul>
+ * <li>{@code nodeId} for {@link FlowNode#getId}
+ * <li>{@code startId} {@link FlowNode#getId} for {@link BlockStartNode}, else {@link BlockEndNode#getStartNode}, else absent
+ * <li>{@code enclosingId} the immediately enclosing {@link BlockStartNode}, if any
+ * <li>{@code label} for {@link LabelAction} if present
+ * </ul>
+ * @see AnnotatedLogAction#annotateHtml
+ */
+@Restricted(NoExternalUse.class)
+public class NewNodeConsoleNote extends ConsoleNote<WorkflowRun> {
+
+    /**
+     * Prefix used in metadata lines.
+     */
+    private static final String CONSOLE_NOTE_PREFIX = "[Pipeline] ";
+
+    public static void print(FlowNode node, TaskListener listener) {
+        try {
+            listener.annotate(new NewNodeConsoleNote(node));
+        } catch (IOException x) {
+            // never mind
+        }
+        listener.getLogger().println(CONSOLE_NOTE_PREFIX + node.getDisplayFunctionName());
+    }
+
+    private final @Nonnull String id;
+    private final @CheckForNull String enclosing;
+    private final @CheckForNull String start;
+
+    private NewNodeConsoleNote(FlowNode node) {
+        id = node.getId();
+        if (node instanceof BlockEndNode) {
+            enclosing = null;
+            start = ((BlockEndNode) node).getStartNode().getId();
+        } else {
+            Filterator<FlowNode> it = FlowScanningUtils.fetchEnclosingBlocks(node).filter(Predicates.not(Predicates.equalTo(node)));
```

Comment:
Future review note: replace with the FlowNode APIs to get enclosing blocks now (much faster lookup).

Reply:
Yeah I could probably make that change easily, good catch.

Reply:
AFAICT `FlowScanningUtils.fetchEnclosingBlocks` just delegates to `FlowNode.iterateEnclosingBlocks`, so the performance will be the same, just avoids a deprecation warning.